src/jou_compiler.h

```diff
@@ -315,14 +317,34 @@ char *signature_to_string(const Signature *sig, bool include_return_type);
 Signature copy_signature(const Signature *sig);
 
 
-// Control Flow Graph.
-// Struct names not prefixed with Cfg because it looks too much like "config" to me
-struct CfVariable {
-    int id;  // Unique, but you can also compare pointers to CfVariable.
-    char name[100];  // Same name as in user's code or empty
+struct Variable {
+    int id;  // Unique, but you can also compare pointers to Variable.
+    char name[100];  // Same name as in user's code, empty for temporary variables created by compiler
     Type type;
     bool is_argument;    // First n variables are always the arguments
 };
+
+struct ExpressionTypes {
+    const AstExpression *expr;
+    Type type;
+    Type *type_after_cast;  // NULL for no implicit cast
+};
+struct TypeContext {
+    const Signature *current_function_signature;
+    // expr_types tells what type each expression has.
+    // It contains nothing for calls to "-> void" functions.
+    List(ExpressionTypes *) expr_types;
+    List(Variable *) variables;
+    List(Type) structs;
+    List(Signature) function_signatures;
+};
+Type type_from_ast(const TypeContext *ctx, const AstType *asttype);
+Type *type_or_void_from_ast(const TypeContext *ctx, const AstType *asttype);
```

Comment:
Maybe there should also be a `signature_from_ast` function. Currently that is done in build_cfg